File: bindings/csharp/src/Psyne/ChannelBuilder.cs
using System;
using Psyne.Native;

namespace Psyne
{
    /// <summary>
    /// Builder class for creating channels with a fluent API.
    /// </summary>
    public sealed class ChannelBuilder
    {
        private string? _uri;
        private long _bufferSize = 1024 * 1024; // 1MB default
        private ChannelMode _mode = ChannelMode.SingleProducerSingleConsumer;
        private ChannelType _type = ChannelType.Single;
        private CompressionConfig? _compression;
        private bool _enableMetrics;

        /// <summary>
        /// Sets the URI for the channel.
        /// </summary>
        /// <param name="uri">The channel URI (e.g., "memory://buffer1", "tcp://localhost:8080").</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WithUri(string uri)
        {
            _uri = uri ?? throw new ArgumentNullException(nameof(uri));
            return this;
        }

        /// <summary>
        /// Sets the buffer size for the channel.
        /// </summary>
        /// <param name="size">The buffer size in bytes.</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WithBufferSize(long size)
        {
            if (size <= 0)
                throw new ArgumentOutOfRangeException(nameof(size), "Buffer size must be positive");
            
            _bufferSize = size;
            return this;
        }

        /// <summary>
        /// Sets the buffer size for the channel using convenient units.
        /// </summary>
        /// <param name="size">The buffer size.</param>
        /// <param name="unit">The size unit (KB, MB, GB).</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WithBufferSize(long size, SizeUnit unit)
        {
            var multiplier = unit switch
            {
                SizeUnit.Bytes => 1L,
                SizeUnit.KB => 1024L,
                SizeUnit.MB => 1024L * 1024L,
                SizeUnit.GB => 1024L * 1024L * 1024L,
                _ => throw new ArgumentOutOfRangeException(nameof(unit))
            };

            return WithBufferSize(size * multiplier);
        }

        /// <summary>
        /// Sets the synchronization mode for the channel.
        /// </summary>
        /// <param name="mode">The channel mode.</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WithMode(ChannelMode mode)
        {
            _mode = mode;
            return this;
        }

        /// <summary>
        /// Sets the channel to single producer, single consumer mode.
        /// </summary>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder SingleProducerSingleConsumer()
        {
            return WithMode(ChannelMode.SingleProducerSingleConsumer);
        }

        /// <summary>
        /// Sets the channel to single producer, multiple consumer mode.
        /// </summary>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder SingleProducerMultipleConsumer()
        {
            return WithMode(ChannelMode.SingleProducerMultipleConsumer);
        }

        /// <summary>
        /// Sets the channel to multiple producer, single consumer mode.
        /// </summary>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder MultipleProducerSingleConsumer()
        {
            return WithMode(ChannelMode.MultipleProducerSingleConsumer);
        }

        /// <summary>
        /// Sets the channel to multiple producer, multiple consumer mode.
        /// </summary>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder MultipleProducerMultipleConsumer()
        {
            return WithMode(ChannelMode.MultipleProducerMultipleConsumer);
        }

        /// <summary>
        /// Sets the channel type.
        /// </summary>
        /// <param name="type">The channel type.</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WithType(ChannelType type)
        {
            _type = type;
            return this;
        }

        /// <summary>
        /// Sets the channel to support a single message type.
        /// </summary>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder SingleType()
        {
            return WithType(ChannelType.Single);
        }

        /// <summary>
        /// Sets the channel to support multiple message types.
        /// </summary>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder MultiType()
        {
            return WithType(ChannelType.Multi);
        }

        /// <summary>
        /// Enables compression with the specified configuration.
        /// </summary>
        /// <param name="compression">The compression configuration.</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WithCompression(CompressionConfig compression)
        {
            _compression = compression ?? throw new ArgumentNullException(nameof(compression));
            return this;
        }

        /// <summary>
        /// Enables LZ4 compression with default settings.
        /// </summary>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WithLz4Compression()
        {
            return WithCompression(CompressionConfig.Lz4());
        }

        /// <summary>
        /// Enables Zstd compression with default settings.
        /// </summary>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WithZstdCompression()
        {
            return WithCompression(CompressionConfig.Zstd());
        }

        /// <summary>
        /// Enables Snappy compression with default settings.
        /// </summary>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WithSnappyCompression()
        {
            return WithCompression(CompressionConfig.Snappy());
        }

        /// <summary>
        /// Enables metrics collection on the channel.
        /// </summary>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WithMetrics()
        {
            _enableMetrics = true;
            return this;
        }

        /// <summary>
        /// Creates a memory-based channel with the specified name.
        /// </summary>
        /// <param name="name">The memory channel name.</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder Memory(string name)
        {
            return WithUri($"memory://{name}");
        }

        /// <summary>
        /// Creates a TCP channel with the specified host and port.
        /// </summary>
        /// <param name="host">The host address.</param>
        /// <param name="port">The port number.</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder Tcp(string host, int port)
        {
            return WithUri($"tcp://{host}:{port}");
        }

        /// <summary>
        /// Creates a Unix socket channel with the specified path.
        /// </summary>
        /// <param name="path">The Unix socket path.</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder UnixSocket(string path)
        {
            return WithUri($"unix://{path}");
        }

        /// <summary>
        /// Creates a UDP multicast channel with the specified address and port.
        /// </summary>
        /// <param name="address">The multicast address.</param>
        /// <param name="port">The port number.</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder UdpMulticast(string address, int port)
        {
            return WithUri($"udp://{address}:{port}");
        }

        /// <summary>
        /// Creates a WebRTC channel for peer-to-peer communication.
        /// </summary>
        /// <param name="peerId">The target peer identifier.</param>
        /// <param name="signalingServerUri">The WebSocket signaling server URI (default: ws://localhost:8080).</param>
        /// <returns>This builder instance for method chaining.</returns>
        public ChannelBuilder WebRtc(string peerId, string signalingServerUri = "ws://localhost:8080")
        {
            return WithUri($"webrtc://{peerId}?signaling={signalingServerUri}");
        }

        /// <summary>
        /// Creates the channel with the configured settings.
        /// </summary>
        /// <returns>A new channel instance.</returns>
        public Channel Build()
        {
            if (string.IsNullOrEmpty(_uri))
                throw new InvalidOperationException("URI must be specified");

            PsyneNative.ErrorCode result;
            IntPtr channelHandle;

            if (_compression != null)
            {
                var nativeCompression = _compression.ToNative();
                result = PsyneNative.psyne_channel_create_compressed(
                    _uri,
                    (UIntPtr)_bufferSize,
                    (PsyneNative.ChannelMode)_mode,
                    (PsyneNative.ChannelType)_type,
                    ref nativeCompression,
                    out channelHandle);
            }
            else
            {
                result = PsyneNative.psyne_channel_create(
                    _uri,
                    (UIntPtr)_bufferSize,
                    (PsyneNative.ChannelMode)_mode,
                    (PsyneNative.ChannelType)_type,
                    out channelHandle);
            }

            PsyneException.ThrowIfError(result);

            var channel = new Channel(channelHandle);

            if (_enableMetrics)
            {
                channel.EnableMetrics();
            }

            return channel;
        }
    }

    /// <summary>
    /// Size units for buffer size specification.
    /// </summary>
    public enum SizeUnit
    {
        /// <summary>Bytes</summary>
        Bytes,
        /// <summary>Kilobytes (1024 bytes)</summary>
        KB,
        /// <summary>Megabytes (1024^2 bytes)</summary>
        MB,
        /// <summary>Gigabytes (1024^3 bytes)</summary>
        GB
    }
}

File: bindings/swift/Sources/Psyne/ChannelBuilder.swift
import Foundation
import CPsyne

/// A builder pattern for creating and configuring channels
public final class ChannelBuilder {
    private var uri: String = ""
    private var bufferSize: Int = 1024 * 1024 // 1MB default
    private var mode: ChannelMode = .spsc
    private var type: ChannelType = .multiType
    private var compression: CompressionConfig? = nil
    
    /// Create a new channel builder
    public init() {}
    
    /// Set the channel to use memory transport
    /// - Parameter name: The name of the memory channel
    /// - Returns: Self for chaining
    public func memory(_ name: String) -> ChannelBuilder {
        self.uri = "memory://\(name)"
        return self
    }
    
    /// Set the channel to use IPC (inter-process communication) transport
    /// - Parameter name: The name of the IPC channel
    /// - Returns: Self for chaining
    public func ipc(_ name: String) -> ChannelBuilder {
        self.uri = "ipc://\(name)"
        return self
    }
    
    /// Set the channel to use TCP transport as a client
    /// - Parameters:
    ///   - host: The host to connect to
    ///   - port: The port to connect to
    /// - Returns: Self for chaining
    public func tcp(host: String, port: Int) -> ChannelBuilder {
        self.uri = "tcp://\(host):\(port)"
        return self
    }
    
    /// Set the channel to use TCP transport as a server
    /// - Parameter port: The port to listen on
    /// - Returns: Self for chaining
    public func tcpServer(port: Int) -> ChannelBuilder {
        self.uri = "tcp://:\(port)"
        return self
    }
    
    /// Set the channel to use Unix domain socket transport as a client
    /// - Parameter path: The path to the Unix socket
    /// - Returns: Self for chaining
    public func unixSocket(_ path: String) -> ChannelBuilder {
        self.uri = "unix://\(path)"
        return self
    }
    
    /// Set the channel to use Unix domain socket transport as a server
    /// - Parameter path: The path to the Unix socket
    /// - Returns: Self for chaining
    public func unixSocketServer(_ path: String) -> ChannelBuilder {
        self.uri = "unix://@\(path)"
        return self
    }
    
    /// Set the channel to use WebSocket transport as a client
    /// - Parameters:
    ///   - host: The host to connect to
    ///   - port: The port to connect to
    /// - Returns: Self for chaining
    public func webSocket(host: String, port: Int) -> ChannelBuilder {
        self.uri = "ws://\(host):\(port)"
        return self
    }
    
    /// Set the channel to use WebSocket transport as a server
    /// - Parameter port: The port to listen on
    /// - Returns: Self for chaining
    public func webSocketServer(port: Int) -> ChannelBuilder {
        self.uri = "ws://:\(port)"
        return self
    }
    
    /// Set the channel to use secure WebSocket transport as a client
    /// - Parameters:
    ///   - host: The host to connect to
    ///   - port: The port to connect to
    /// - Returns: Self for chaining
    public func secureWebSocket(host: String, port: Int) -> ChannelBuilder {
        self.uri = "wss://\(host):\(port)"
        return self
    }
    
    // ===================================================================
    // v1.3.0 Transport Methods
    // ===================================================================
    
    /// Set the channel to use UDP multicast transport
    /// - Parameters:
    ///   - multicastAddress: The multicast group address (e.g., "239.255.0.1")
    ///   - port: The port number
    /// - Returns: Self for chaining
    public func multicast(multicastAddress: String, port: Int) -> ChannelBuilder {
        self.uri = "udp://\(multicastAddress):\(port)"
        return self
    }
    
    /// Set the channel to use WebRTC transport for peer-to-peer communication
    /// - Parameters:
    ///   - peerId: The target peer identifier
    ///   - signalingServerUri: The WebSocket signaling server URI (default: ws://localhost:8080)
    /// - Returns: Self for chaining
    public func webrtc(peerId: String, signalingServerUri: String = "ws://localhost:8080") -> ChannelBuilder {
        self.uri = "webrtc://\(peerId)?signaling=\(signalingServerUri)"
        return self
    }
    
    /// Set a custom URI for the channel
    /// - Parameter uri: The custom URI
    /// - Returns: Self for chaining
    public func withUri(_ uri: String) -> ChannelBuilder {
        self.uri = uri
        return self
    }
    
    /// Set the buffer size for the channel
    /// - Parameter size: The buffer size in bytes
    /// - Returns: Self for chaining
    public func withBufferSize(_ size: Int) -> ChannelBuilder {
        self.bufferSize = size
        return self
    }
    
    /// Set the channel mode (threading configuration)
    /// - Parameter mode: The channel mode
    /// - Returns: Self for chaining
    public func withMode(_ mode: ChannelMode) -> ChannelBuilder {
        self.mode = mode
        return self
    }
    
    /// Set the channel type (single or multi-type messages)
    /// - Parameter type: The channel type
    /// - Returns: Self for chaining
    public func withType(_ type: ChannelType) -> ChannelBuilder {
        self.type = type
        return self
    }
    
    /// Set compression configuration for the channel
    /// - Parameter config: The compression configuration
    /// - Returns: Self for chaining
    public func withCompression(_ config: CompressionConfig) -> ChannelBuilder {
        self.compression = config
        return self
    }
    
    /// Enable LZ4 compression with default settings
    /// - Returns: Self for chaining
    public func withLZ4Compression() -> ChannelBuilder {
        self.compression = .lz4Fast
        return self
    }
    
    /// Enable Zstandard compression with high compression ratio
    /// - Returns: Self for chaining
    public func withZstdCompression() -> ChannelBuilder {
        self.compression = .zstdHigh
        return self
    }
    
    /// Enable Snappy compression with balanced performance
    /// - Returns: Self for chaining
    public func withSnappyCompression() -> ChannelBuilder {
        self.compression = .snappy
        return self
    }
    
    /// Configure for single producer, single consumer (highest performance)
    /// - Returns: Self for chaining
    public func singleProducerSingleConsumer() -> ChannelBuilder {
        self.mode = .spsc
        return self
    }
    
    /// Configure for single producer, multiple consumers
    /// - Returns: Self for chaining
    public func singleProducerMultipleConsumers() -> ChannelBuilder {
        self.mode = .spmc
        return self
    }
    
    /// Configure for multiple producers, single consumer
    /// - Returns: Self for chaining
    public func multipleProducersSingleConsumer() -> ChannelBuilder {
        self.mode = .mpsc
        return self
    }
    
    /// Configure for multiple producers, multiple consumers
    /// - Returns: Self for chaining
    public func multipleProducersMultipleConsumers() -> ChannelBuilder {
        self.mode = .mpmc
        return self
    }
    
    /// Configure for single message type (optimized)
    /// - Returns: Self for chaining
    public func singleMessageType() -> ChannelBuilder {
        self.type = .singleType
        return self
    }
    
    /// Configure for multiple message types
    /// - Returns: Self for chaining
    public func multipleMessageTypes() -> ChannelBuilder {
        self.type = .multiType
        return self
    }
    
    /// Build the channel with the configured settings
    /// - Returns: A new `Channel` instance
    /// - Throws: `PsyneError` if channel creation fails
    public func build() throws -> Channel {
        guard !uri.isEmpty else {
            throw PsyneError.invalidArgument("Channel URI must be set")
        }
        
        guard bufferSize > 0 else {
            throw PsyneError.invalidArgument("Buffer size must be greater than 0")
        }
        
        var channelHandle: UnsafeMutablePointer<psyne_channel_t>?
        let result: psyne_error_t
        
        if let compression = compression {
            var compressionConfig = compression.cValue
            result = psyne_channel_create_compressed(
                uri,
                bufferSize,
                mode.cValue,
                type.cValue,
                &compressionConfig,
                &channelHandle
            )
        } else {
            result = psyne_channel_create(
                uri,
                bufferSize,
                mode.cValue,
                type.cValue,
                &channelHandle
            )
        }
        
        try throwOnError(result, context: "Failed to create channel with URI: \(uri)")
        
        guard let handle = channelHandle else {
            throw PsyneError.unknown(-1)
        }
        
        return Channel(handle: handle)
    }
    
    /// Get a description of the current configuration
    public var description: String {
        var components = [String]()
        
        if !uri.isEmpty {
            components.append("URI: \(uri)")
        }
        
        components.append("Buffer: \(bufferSize) bytes")
        components.append("Mode: \(mode)")
        components.append("Type: \(type)")
        
        if let compression = compression {
            components.append("Compression: \(compression.type)")
        }
        
        return "ChannelBuilder(\(components.joined(separator: ", ")))"
    }
}